WasatchNET/EEPROM.cs

```diff
@@ -1030,10 +1033,12 @@ public bool write()
 
                 if (subformat == PAGE_SUBFORMAT.USER_DATA)
                 {
-                    Array.Copy(userData, 64, pages[6], 0, 64);
-                    Array.Copy(userData, 128, pages[7], 0, 64);
+                    if (userData.Length >= 128)

```

Comment:
I don't think these length checks are right.  Instead of 128 and 192, shouldn't they be 64 and 128?

Reply:
I'm accessing the byte array in chunks of 64, so I think it does have to be like this? I'm not positive what will happen if Copy tries to read byte 66 if there are only 65 bytes. I can try to make a test module in WPSC. Could also add guards in the setter to enforce length?

Reply:
So this seems to be what you have now:

```
if (userData.Length >= 128)
  Array.Copy(userData, 64, pages[6], 0, 64);
if (userData.Length >= 192)
  Array.Copy(userData, 128, pages[7], 0, 64);
```

That doesn't make sense to me in two ways.  First of all, since the pages are 64 bytes apiece, it seems like it should be closer to this:

```
if (userData.Length >= 64) // greater than one page, so write second
  Array.Copy(userData, 64, pages[6], 0, 64);
if (userData.Length >= 128) // greater than two pages, so write third
  Array.Copy(userData, 128, pages[7], 0, 64);
```

But even there, I think the comparison operators are still wrong, because a length of 64 bytes will actually fit the first page, and a length of 128 bytes will fit the second page.  Therefore, this seems like the proper check:

```
if (userData.Length >= 65) // greater than one page, so write second
  Array.Copy(userData, 64, pages[6], 0, 64);
if (userData.Length >= 129) // greater than two pages, so write third
  Array.Copy(userData, 128, pages[7], 0, 64);
```

Maybe we're talking past each other?  In any case, I suggest you actually test it with different sizes of userData (30 char, 64 char, 90 char, 128 char, 150 char, 192 char, 200 char) to confirm the actual behavior.

Reply:
I do think we are talking past each other: you're concerned about data that's there being lost and I'm concerned about trying to access data that isn't there. I think I can make this code smarter anyway, so I think I will do that then push it out then try out some testing